File: python/antikythera/session.py
"""Session management implementation."""

from __future__ import annotations

import time
from typing import Optional

from antikythera.types import SessionInfo


class SessionManager:
    """Session lifecycle management.

    Automatically manages session creation, TTL, and cleanup.

    Example:
        >>> manager = SessionManager(max_sessions=100)
        >>> session = manager.get_or_create("session-123", "coder")
        >>> print(session.session_id)
    """

    def __init__(
        self,
        max_sessions: int = 100,
        session_ttl_ms: int = 3600000,
    ):
        """Create a new SessionManager.

        Args:
            max_sessions: Maximum concurrent sessions.
            session_ttl_ms: Session time-to-live in milliseconds.
        """
        self._max_sessions = max_sessions
        self._ttl_ms = session_ttl_ms
        self._sessions: dict[str, SessionInfo] = {}

    def get_or_create(
        self, session_id: str, agent_id: str
    ) -> SessionInfo:
        """Get or create a session.

        Args:
            session_id: Session identifier.
            agent_id: Agent identifier.

        Returns:
            Session information.
        """
        now = int(time.time() * 1000)

        # Return existing session if found
        if session_id in self._sessions:
            session = self._sessions[session_id]
            session.last_activity = now
            return session

        # Evict if at capacity
        if len(self._sessions) >= self._max_sessions:
            self._evict_expired()
            if len(self._sessions) >= self._max_sessions:
                # Evict oldest
                oldest = min(
                    self._sessions.values(), key=lambda s: s.last_activity
                )
                del self._sessions[oldest.session_id]

        # Create new session
        session = SessionInfo(
            session_id=session_id,
            agent_id=agent_id,
            created_at=now,
            last_activity=now,
            message_count=0,
        )
        self._sessions[session_id] = session
        return session

    def get(self, session_id: str) -> Optional[SessionInfo]:
        """Get a session by ID.

        Args:
            session_id: Session identifier.

        Returns:
            Session information or None.
        """
        return self._sessions.get(session_id)

    def list_by_agent(self, agent_id: str) -> list[SessionInfo]:
        """List sessions for an agent.

        Args:
            agent_id: Agent identifier.

        Returns:
            List of sessions.
        """
        return [s for s in self._sessions.values() if s.agent_id == agent_id]

    def list_all(self) -> list[SessionInfo]:
        """List all sessions.

        Returns:
            All sessions.
        """
        return list(self._sessions.values())

    def remove(self, session_id: str) -> Optional[SessionInfo]:
        """Remove a session.

        Args:
            session_id: Session identifier.

        Returns:
            Removed session or None.
        """
        return self._sessions.pop(session_id, None)

    def count(self) -> int:
        """Get session count.

        Returns:
            Number of active sessions.
        """
        return len(self._sessions)

    def _evict_expired(self) -> None:
        """Evict expired sessions."""
        now = int(time.time() * 1000)
        expired = [
            sid
            for sid, session in self._sessions.items()
            if now - session.last_activity > self._ttl_ms
        ]
        for sid in expired:
            del self._sessions[sid]

File: antikythera-core/src/application/agent/multi_agent/session.rs
//! Multi-session management for the orchestrator.
//!
//! Tracks sessions across agents, supports concurrent sessions,
//! and provides lifecycle management (creation, activity tracking,
//! TTL-based eviction).

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use crate::logging::{OrchestratorLogger, SessionContext};

/// Session state tracked by the orchestrator.
#[derive(Debug, Clone)]
pub struct ManagedSession {
    pub session_id: String,
    pub agent_id: String,
    pub created_at: Instant,
    pub last_activity: Instant,
    pub message_count: usize,
    pub metadata: HashMap<String, String>,
}

impl ManagedSession {
    /// Return how long since the last activity.
    pub fn idle_duration(&self) -> Duration {
        self.last_activity.elapsed()
    }

    /// Return how long since creation.
    pub fn age(&self) -> Duration {
        self.created_at.elapsed()
    }
}

/// Manages multiple sessions across agents.
///
/// Thread-safe via `Arc<Mutex<...>>` — cheap to clone and pass into async
/// closures.
#[derive(Clone)]
pub struct OrchestratorSessionManager {
    sessions: Arc<Mutex<HashMap<String, ManagedSession>>>,
    max_sessions: usize,
    session_ttl: Option<Duration>,
    log: OrchestratorLogger,
}

impl OrchestratorSessionManager {
    /// Create a new session manager with default limits.
    pub fn new() -> Self {
        Self {
            sessions: Arc::new(Mutex::new(HashMap::new())),
            max_sessions: 100,
            session_ttl: None,
            log: OrchestratorLogger::new(&SessionContext::default().into_session_id()),
        }
    }

    /// Set maximum concurrent sessions.
    pub fn with_max_sessions(mut self, max: usize) -> Self {
        self.max_sessions = max;
        self
    }

    /// Set session time-to-live. Sessions idle longer than this are evicted
    /// when the session limit is reached.
    pub fn with_session_ttl(mut self, ttl: Duration) -> Self {
        self.session_ttl = Some(ttl);
        self
    }

    /// Create or get a session for an agent.
    ///
    /// If a session with `session_id` already exists, its `last_activity` is
    /// refreshed and it is returned. Otherwise a new session is created.
    pub fn get_or_create(
        &self,
        session_id: &str,
        agent_id: &str,
    ) -> Result<ManagedSession, SessionError> {
        let mut sessions = self.sessions.lock().unwrap();

        // Existing session — refresh activity.
        if let Some(session) = sessions.get(session_id) {
            let mut session = session.clone();
            session.last_activity = Instant::now();
            sessions.insert(session_id.to_string(), session.clone());
            self.log
                .debug(format!("Session refreshed | session_id={}", session_id));
            return Ok(session);
        }

        // Capacity check — try eviction first.
        if sessions.len() >= self.max_sessions {
            self.evict_expired(&mut sessions);
            if sessions.len() >= self.max_sessions {
                self.log.warn(format!(
                    "Session limit reached | max={} current={}",
                    self.max_sessions,
                    sessions.len()
                ));
                return Err(SessionError::MaxSessionsReached(self.max_sessions));
            }
        }

        // Create new session.
        let session = ManagedSession {
            session_id: session_id.to_string(),
            agent_id: agent_id.to_string(),
            created_at: Instant::now(),
            last_activity: Instant::now(),
            message_count: 0,
            metadata: HashMap::new(),
        };
        sessions.insert(session_id.to_string(), session.clone());
        self.log.debug(format!(
            "Session created | session_id={} agent_id={}",
            session_id, agent_id
        ));
        Ok(session)
    }

    /// Record a message exchange in a session (bumps `message_count` and
    /// `last_activity`).
    pub fn record_message(&self, session_id: &str) {
        let mut sessions = self.sessions.lock().unwrap();
        if let Some(session) = sessions.get_mut(session_id) {
            session.message_count += 1;
            session.last_activity = Instant::now();
        }
    }

    /// Get a snapshot of a session by ID.
    pub fn get(&self, session_id: &str) -> Option<ManagedSession> {
        self.sessions.lock().unwrap().get(session_id).cloned()
    }

    /// List all sessions for a given agent.
    pub fn list_by_agent(&self, agent_id: &str) -> Vec<ManagedSession> {
        self.sessions
            .lock()
            .unwrap()
            .values()
            .filter(|s| s.agent_id == agent_id)
            .cloned()
            .collect()
    }

    /// List all tracked sessions.
    pub fn list_all(&self) -> Vec<ManagedSession> {
        self.sessions.lock().unwrap().values().cloned().collect()
    }

    /// Remove a session, returning it if it existed.
    pub fn remove(&self, session_id: &str) -> Option<ManagedSession> {
        let removed = self.sessions.lock().unwrap().remove(session_id);
        if removed.is_some() {
            self.log
                .debug(format!("Session removed | session_id={}", session_id));
        }
        removed
    }

    /// Return total tracked session count.
    pub fn count(&self) -> usize {
        self.sessions.lock().unwrap().len()
    }

    /// Evict sessions whose `last_activity` exceeds the configured TTL.
    fn evict_expired(&self, sessions: &mut HashMap<String, ManagedSession>) {
        if let Some(ttl) = self.session_ttl {
            let before = sessions.len();
            sessions.retain(|_, session| session.last_activity.elapsed() < ttl);
            let evicted = before - sessions.len();
            if evicted > 0 {
                self.log
                    .debug(format!("Expired sessions evicted | count={}", evicted));
            }
        }
    }
}

impl Default for OrchestratorSessionManager {
    fn default() -> Self {
        Self::new()
    }
}

/// Errors produced by session management operations.
#[derive(Debug, Clone, thiserror::Error)]
pub enum SessionError {
    #[error("maximum concurrent sessions reached ({0})")]
    MaxSessionsReached(usize),
    #[error("session not found: {0}")]
    SessionNotFound(String),
}
